packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -85,8 +85,12 @@ export function FocusScope(props: FocusScopeProps) {
   let endRef = useRef<HTMLSpanElement>();
   let scopeRef = useRef<Element[]>([]);
   let ctx = useContext(FocusContext);
-  // if there is no scopeRef on the context, then the parent is the focusScopeTree's root, represented by null
-  let parentScope = ctx?.scopeRef ?? null;
+
+  // The parent scope is based on the JSX tree, using context.
+  // However, if a new scope mounts outside the active scope (e.g. DialogContainer launched from a menu),
+  // we want the parent scope to be the active scope instead.
+  let ctxParent = ctx?.scopeRef ?? null;
+  let parentScope = useMemo(() => activeScope && focusScopeTree.getTreeNode(activeScope) && !isAncestorScope(activeScope, ctxParent) ? activeScope : ctxParent, [ctxParent]);
```

Comment:
Trying to understand the logic here:
- for DialogContainer from menu case, the DialogContainer's FocusScope ctxParent would be null since it is outside the Menu's JSX tree. The parent scope is thus set to the active scope, aka the Menu's FocusScope
- for a nested DialogTrigger case, a nested Dialog's FocusScope ctxParent is its parent Dialog since it is a child in the JSX tree. The activeScope is equal to the ctxParent (aka both are the parent Dialog) and thus we set the parentScope as the activeScope since isAncestorScope still resolves as false
- for a newly mounted FocusScope, activeScope is null so we set parentScope to null

What about when a sibling focus scope mounts when the first sibling scope is active? Wouldn't this evaluate in such a way that parentScope of the second sibling is set to the first sibling scope?

Reply:
It would, yes. This is essentially the DialogContainer + Menu case. For our purposes, I think this is ok. The user would expect focus to return to the previously active scope when the new one unmounts.

Actually, in most cases, the parent should really be the active scope at the time the new scope mounts. The case where this is not true is if multiple nested scopes mount at the same time.